consul_exporter.go

```diff
@@ -0,0 +1,216 @@
+package main
+
+import (
+	"flag"
+	_ "fmt"
+	_ "io"
+	_ "io/ioutil"
+	"log"
+	_ "net"
+	"net/http"
+	_ "strconv"
+	_ "strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "consul"
+)
+
+var (
+	serviceLabelNames = []string{"service", "node"}
+	memberLabelNames  = []string{"member"}
+)
+
+// Exporter collects HAProxy stats from the given URI and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	URI   string
+	mutex sync.RWMutex
+
+	up, clusterServers              prometheus.Gauge
+	totalQueries, jsonParseFailures prometheus.Counter
+	serviceMetrics, lockMetrics     map[string]*prometheus.GaugeVec
+	client                          *api.Client
+}
+
+func newServiceMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "service_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		serviceLabelNames,
+	)
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(uri string, consulLocks string, timeout time.Duration) *Exporter {
+	// connect to Consul
+
+	consul_client, _ := api.NewClient(&api.Config{
+		Address: uri,
+	})
+
+	// init our exporter
+
+	return &Exporter{
+		URI: uri,
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "up",
+			Help:      "Was the last query of Consul successful.",
+		}),
+		totalQueries: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_total_queries",
```

Comment:
Actually, this confused me. I thought it meant the number of client queries against Consul, not how often this exporter has queried Consul. Yeah, I think this can be removed. But stats about client queries could actually be useful, if they're obtainable.